### PR TITLE
[#98] Supply endpoints

### LINKED ISSUE #98: Add "Supply" endpoints
As CMC requires an API to fetch supply data for blockchains, a new endpoint should be created

- circulating supply (take unlock excel file, burn address(es) and on-chain burn into account)
- max supply

CMC requires the output of the API to be the raw number including 8 decimals.

example HTTP response:
```
HTTP/1.1 200 OK
Date: Tue, 06 Jul 2021 14:29:29 GMT
Content-Type: text/plain

77908483.88736623
```

a raw `77908483.88736623` should be returned.

Perhaps make a default with JSON and make it raw if ?output=raw is applied as URL parameter.
example: `http://chainz.cryptoid.info/grs/api.dws?q=totalcoins`

### PR DESCRIPTION
**Pull request by brunodb3**
- Adds `/supply/circulating` endpoint
  - GET for returning the circulating supply
    - total supply (LTO + LTO20)
    - minus burned (LTO + LTO20)
    - minus transaction fee burn
    - minus locked supply based on spreadsheet data
  - Adds indexing of transaction fee burn for quicker calculation
- Adds `/supply/max` endpoint
  - GET for returning the max supply
    - total supply (LTO + LTO20)

---

closes #98

### CODE REVIEW

**Review by KaloyanTanev — CHANGES_REQUESTED** [review]

File: src/supply/locked-supply.data.json [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  },
+  "2021": {
+    "jan": 294114776,
+    "feb": 294156442,
+    "mar": 298198109,
+    "apr": 306289776,
+    "may": 310331442,
+    "jun": 314373109,
+    "jul": 322464776,
+    "aug": 326506442,
```
[the remark this answers is not in the document]
Comment (brunodb3) [reply src/supply/locked-supply.data.json:38]: This is data extracted from the "lock" spreadsheet. It's static data for amount of locked tokens on each month

https://docs.google.com/spreadsheets/d/1SaPvNOHk8SPMEUP9L-1OOBpGw8b6xKxQkmJE9f5zsA8/edit#gid=604296154

File: src/supply/supply.controller.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  constructor(
+    private readonly logger: LoggerService,
+    private readonly supplyService: SupplyService,
+  ) { }
+
+  @Get('circulating')
+  @ApiOperation({ summary: 'Get circulating supply' })
+  @ApiQuery({ name: 'output', required: false, description: 'Flag for type of output', enum: ['raw'] })
+  @ApiResponse({ status: 200 })
+  @ApiResponse({ status: 500, description: 'failed to get circulating supply' })
```
[the remark this answers is not in the document]
Comment (brunodb3) [reply src/supply/supply.controller.ts:19]: I'm just keeping it consistent with the other endpoints. Plus, a 5xx means something happened server side, which would be the case for this particular error

File: src/supply/supply.controller.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      return res.status(500).json({ error: `failed to get circulating supply: ${error}` });
+    }
+
+  }
+
+  @Get('max')
+  @ApiOperation({ summary: 'Get max supply' })
+  @ApiQuery({ name: 'output', required: false, description: 'Flag for type of output', enum: ['raw'] })
+  @ApiResponse({ status: 200 })
+  @ApiResponse({ status: 500, description: 'failed to get max supply' })
```
[the remark this answers is not in the document]
Comment (brunodb3) [reply src/supply/supply.controller.ts:41]: Same answer :)

File: src/supply/supply.service.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    return this.fixedDecimals(result);
+  }
+
+  async getMaxSupply(): Promise<string> {
+    const bridgeStats = await this.request.get<BridgeStats>(`${this.bridgeUrl}/stats`);
+    if (bridgeStats instanceof Error) {
+      return Promise.reject(bridgeStats);
+    }
+
+    const maxSupply = bridgeStats.data.volume.lto.supply + bridgeStats.data.volume.lto20.supply;
```
[the remark this answers is not in the document]
Comment (brunodb3) [reply src/supply/supply.service.ts:73]: Actually not sure on this. It makes sense, I'll ask Arnold

File: src/supply/supply.service.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+    return this.fixedDecimals(maxSupply)
+  }
+
+  private calculateCirculatingSupply(stats: BridgeStats, locked: number, burnFee: number): number {
+    return (
```
[the remark this answers is not in the document]
Comment (brunodb3) [reply src/supply/supply.service.ts:79]: Yup, same answer :)

File: src/supply/supply.service.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  async incrTxFeeBurned(blockHeight: number): Promise<void> {
+    const isFeatureActivated = await this.isFeatureActivated(blockHeight);
+    if (!isFeatureActivated) {
+      return;
+    }
+
+    const current = await this.storage.getTxFeeBurned().catch<number>(() => 0);
+
+    const burnFee = 0.1;
+    const newValue = current + (1 * burnFee);
```
[the remark this answers is not in the document]
Comment (brunodb3) [reply src/supply/supply.service.ts:102]: We don't, I'm a dummy :)

File: src/supply/supply.controller.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (KaloyanTanev) [reply src/supply/supply.controller.ts:19]: If on the other places is like that, yeah, leave it. I find it a bit weird that the actual error's code is neglected and substituted with 500.

File: src/supply/supply.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (brunodb3) [reply src/supply/supply.service.ts]: Updated to include `binance` supply and fixed the calculation :) (`supply` already takes `burned` into consideration)

**Review by KaloyanTanev — APPROVED** [review]

**Pull request merged**